ttnn/cpp/ttnn/operations/data_movement.hpp

```diff
@@ -315,9 +319,114 @@ struct RepeatInterleave {
+    }
+
+    template <typename... Args>
+    static auto input_tensors_to_validate(const ttnn::Tensor& input_tensor, Args&&... args) {
+        return std::make_tuple(input_tensor);
+    }
+
+    static ttnn::Tensor execute_on_worker_thread(
+        const ttnn::Tensor& input_tensor,
+        std::vector<std::pair<uint32_t, uint32_t>> padding, //intentionally not const&
+        const float value,
+        const std::optional<MemoryConfig>& memory_config_arg) {
+
+        const int original_rank = input_tensor.get_shape().rank();
+        if(int diff = original_rank - padding.size(); diff != 0) {
+            TT_FATAL(diff > 0, "ttnn.pad: padding len can't be larger than input tensor rank");
+
+            padding.insert(padding.begin(), diff, {0, 0});
+        }
+
+        TT_FATAL(
+            padding.size() == original_rank,
+            "ttnn.pad: padding must be the same length as the input tensor rank");
+        TT_FATAL(
+            input_tensor.get_layout() != ttnn::ROW_MAJOR_LAYOUT,
+            "ttnn.pad: row-major tensors have to use fallback because the kernel currently causes a PCC error");
+
+        // Unsqueeze Tensor to 4D if it is not already
+        ttnn::Tensor input_tensor_4D = ttnn::unsqueeze_to_4D(input_tensor);        
+        padding.insert(padding.begin(), 4 - original_rank, {0, 0});
+        auto input_shape_with_tile_padding = input_tensor_4D.get_shape().with_tile_padding();
+        std::vector<uint32_t> output_padded_shape(padding.size());
+        for(size_t i = 0; i < padding.size(); i++) {
+            output_padded_shape[i] = input_shape_with_tile_padding[i] + padding[i].second;
+        }
+
+        // Due to the strangeness of tt::tt_metal::pad, we need to split front and back pad
+        // Front will be passed separately. And pad_back is retrieved -> output_padded_shape - pad_front
+        auto memory_config = memory_config_arg.value_or(input_tensor.memory_config());
+        auto pad_front = padding | std::views::transform([](const auto& p) { return p.first; });
+        auto pad_back = padding | std::views::transform([](const auto& p) { return p.second; });
+
+        const bool front_padding_is_zero = std::accumulate(pad_front.begin(), pad_front.end(), 0) == 0;
+        TT_FATAL(
+            front_padding_is_zero,
+            "ttnn.pad: on device padding does not support front padding");
+
+        const int target_height = output_padded_shape[output_padded_shape.size() - 2];
+        const int target_width = output_padded_shape[output_padded_shape.size() - 1];
+        TT_FATAL(
+            target_height % ttnn::TILE_SIZE == 0 || target_width % ttnn::TILE_SIZE == 0,
+            "ttnn.pad: for tiled tensors padding end must be a multiple of the tile size on height and width for a "
+            "tensor in tile layout");
+
+        // Performing actual padding
+        // std::vector<uint32_t> pad_front_vec;
+        // std::ranges::copy(pad_front, std::back_inserter(pad_front_vec));
```

Comment:
are you planning on switching this back to ranges now that we have the standard lib support?

Reply:
No need. But I should cleanup this comment, it is outdated.
The rest of the code actively uses ranges/views as I made this PR with full C++20 locally with newer stdlibc++.